### PR TITLE
[receiver/saphanareceiver] New SAP HANA metrics receiver

### PR DESCRIPTION
**Pull request by dehaansa**
**Description:**
Adds a new SAP HANA metrics receiver.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/8827

**Testing:** Local testing & unit testing. Integration tests were attempted, but could not get the [sap hana test docker image](https://hub.docker.com/_/sap-hana-express-edition/plans/f2dc436a-d851-4c22-a2ba-9de07db7a9ac?tab=instructions) running within testcontainers.

### CODE REVIEW

**Comment by dehaansa** [comment]
Current unit test failure is `--- FAIL: TestEndToEndSummarySupport (5.11s)` for `prometheusexporter`, shouldn't be related to this PR.

File: receiver/saphanareceiver/metadata.yaml
Comment (djaglowski) [inline receiver/saphanareceiver/metadata.yaml:19]: Thoughts on which of these should be resource attributes?

Comment (pmm-sumo) [reply receiver/saphanareceiver/metadata.yaml:19]: If that helps, wanted to bring that we have resource conventions for [host](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/host.md), among others

At the same, we do not have database semantic conventions for metrics, but we [have those for traces](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md) and I believe the idea is to be consistent across the signals, if possible. There's `db.system` and few others listed there

Comment (dehaansa) [reply receiver/saphanareceiver/metadata.yaml:19]: I'll look at refactoring to collect resource attributes. I'm concerned that it's going to add significant complexity to the receiver, but I'll certainly take a look.

`db.system` in particular doesn't mean the same thing in SAP HANA (an identifier of each hana database) as it appears to in traces (the type of database system like "sap hana" or "mysql") so I would hesitate to try to align this attribute to that trace attribute.

Comment (pmm-sumo) [reply receiver/saphanareceiver/metadata.yaml:19]: Perhaps "system" is closer to instance name or database name then? Those are [mentioned here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md#connection-level-attributes-for-specific-technologies)

Comment (pmm-sumo) [reply receiver/saphanareceiver/metadata.yaml:19]: I'm looking at the other database receivers (mysqlreceiver, mongodbreceiver, postgresqlreceiver) and I think they are actually not following the specification (or, the intention of it). The problem is that we currently lack some of the metric-related attributes but there seems to be agreement that all signals should use [coherent attribute names](https://github.com/open-telemetry/opentelemetry-specification/issues/2488)

I think we have a couple of issues here:
1. Some of the attributes should be resource-level attributes - I think most if not all of the above (host, database, name, port, etc.) fit resource better than record
2. We should do better job at fitting existing attribute names into new receivers - e.g. I still think `db.system` should be populated with `saphana` or so. This will be still helpful when looking at metrics of different database systems 
3. Whenever technology-specific attribute is being used, it should be actually namespaced (so e.g. `saphana.primary_host`) 

As a counterexample, [sqlserverreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/sqlserverreceiver/metadata.yaml) and [couchdbreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/couchdbreceiver/metadata.yaml) do a better job here

Comment (dehaansa) [reply receiver/saphanareceiver/metadata.yaml]: @pmm-sumo 
1. I've moved up "host" so far. Moving many of the other attributes to resource level would lead to a lot of resources with 1 or 2 metrics, and a few resources with many many metrics. Is moving up attributes to the resource level still desirable when it ends up fragmenting the collection like that?
2. Added db.system
3. Is this specifically when the attributes are resource level? At the metric level it seems very redundant as the metric names are already namespaced. SQL Server & CouchDB that you linked to only appear to follow that on their one resource level attribute.

Comment (pmm-sumo) [reply receiver/saphanareceiver/metadata.yaml]: Thank you @dehaansa This are good points

Ad. 1 Thanks! I think that the fragmentation is uncommon and indication that attribute should be not at the resource level. Out of the above, do you have some examples of attributes which would cause it? 

Ad. 3 We have [recommendations on naming attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md). Actually it has two parts, one for [OpenTelemetry Authors](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md#recommendations-for-opentelemetry-authors) and the other for [Application Developers](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md#recommendations-for-application-developers). The spec uses "SHOULD" on using namespaces. The practice is not strictly followed. I guess that this is worth bringing to the Spec SIG and opening an issue (I will take care of that)

Comment (dehaansa) [reply receiver/saphanareceiver/metadata.yaml]: 1. `database` attribute is only used on the saphana.uptime metric, for example. `system` is used on saphana.uptime and also on the (3) saphana.license.* metrics along with `product`. `primary`/`secondary`/`port`/`mode` are only used on the (again 3) saphana.replication.* metrics. Those metrics are also all not associated with a single SAP HANA host, so they are currently on a "resource" without any attributes other than the `db.system`.

3. If there's value here I'm happy to update them, will wait to see at least initial conversation on that spec issue! Thanks for opening it.

Comment (pmm-sumo) [reply receiver/saphanareceiver/metadata.yaml]: There wasn't much additional discussion on the Attributes naming beyound the SIG. I'm not going to block the review due to that. If we'll make one decision I guess we'll then just modify the other receivers.

As for the resource attributes, I agree that `database` is not a good fit (since, I believe, there might be several "databases" for a single SAP HANA instance?). I'm not so sure about `product` (is it not relevant for other metrics?) and similarly about the `primary`/`secondary` and so on. Though ultimately, I trust your judgement as SAP HANA SME :). I'm fine with leaving them at metric level if they are a better fit there

Comment (dehaansa) [reply receiver/saphanareceiver/metadata.yaml]: Multiple databases is possible, yes. `product` is a very license specific value, not necessarily relevant to the other metrics directly, and I believe there's certain edge cases where multiple licenses can coexist. The other ones (`primary`/`secondary`) are specific to the different modes of replication so there can be multiple depending on the size of the cluster and not directly relevant to the other metrics.

I appreciate your thoroughness on this review!

File: receiver/saphanareceiver/client.go
Comment (pmm-sumo) [inline receiver/saphanareceiver/client.go:164]: Perhaps instead of iterating here, as well `orderedMetricLabels` and `orderedStats` below, rowFields could be created with single make? (since we know the sizes of each)

File: receiver/saphanareceiver/factory.go
Comment (pmm-sumo) [inline receiver/saphanareceiver/factory.go:49]: A nit, but I think it's a nice practice when default values are represented as constant values

File: receiver/saphanareceiver/metadata.yaml
Comment (pmm-sumo) [inline receiver/saphanareceiver/metadata.yaml:3]: To avoid clash with [host semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/host.md) I think we might use `saphana.host` here maybe, or move it back to metrics attributes. Alternatively `net.peer.name` (as is mentioned in [tracing database semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md)) might be used - though probably it's more fit for the metric attribute

**Comment by pmm-sumo** [comment]
Some nits, it looks good overall. I will make another run tomorrow

File: receiver/saphanareceiver/client.go
Comment (dehaansa) [reply receiver/saphanareceiver/client.go]: rowFields needs to be a set of pointers to objects, so it needs more than just a singular make in my understanding, but I was able to update it to be a little bit cleaner.

**Review by pmm-sumo — APPROVED** [review]
This is a really large PR and I'm sure it required a fair amount of work. Thank you for the contribution! I think it looks good overall. Another pair of eyes would be helpful before merging this

File: CHANGELOG.md
Comment (pmm-sumo) [inline CHANGELOG.md:41]: This should go to line 13 or so

**Review by djaglowski — APPROVED** [review]

**Pull request merged**
